Don't add non-integer Triton kernel arg 1 to equal_to_1

aakhundov opened the pull request:
Stack from [ghstack](https://github.com/ezyang/ghstack) (oldest at bottom):
* __->__ #123886
* #123703

Summary: Triton compiler adds constnat argument 1 to `equal_to_1` [only when it's an int](https://github.com/openai/triton/blob/8c5e33c77ef83e0cb99c744e58842930e602df31/python/triton/runtime/jit.py#L275). Here we restrict Inductor's `equal_to_1` in the same way.

Test Plan:

```
$ python test/inductor/test_triton_kernels.py -k test_triton_kernel_equal_to_1_float_arg
...
----------------------------------------------------------------------
Ran 1 test in 6.528s

OK

$ python test/inductor/test_triton_kernels.py -k test_triton_kernel_equal_to_1_arg
...
----------------------------------------------------------------------
Ran 2 tests in 10.142s

OK
```

cc @voznesenskym @penguinwu @EikanWang @jgong5 @Guobing-Chen @XiaobingSuper @zhuhaozhe @blzheng @wenzhe-nrv @jiayisunx @peterbell10 @ipiszy @yf225 @chenyang78 @kadeng @muchulee8 @ColinPeppler @amjames @desertfire @chauhang

aakhundov committed:
Don't add non-integer Triton kernel arg 1 to equal_to_1

Summary: Triton compiler adds constnat argument 1 to `equal_to_1` [only when it's an int](https://github.com/openai/triton/blob/8c5e33c77ef83e0cb99c744e58842930e602df31/python/triton/runtime/jit.py#L275). Here we restrict Inductor's `equal_to_1` in the same way.

Test Plan:

```
$ python test/inductor/test_triton_kernels.py -k test_triton_kernel_equal_to_1_float_arg
...
----------------------------------------------------------------------
Ran 1 test in 6.528s

OK

$ python test/inductor/test_triton_kernels.py -k test_triton_kernel_equal_to_1_arg
...
----------------------------------------------------------------------
Ran 2 tests in 10.142s

OK
```

[ghstack-poisoned]

oulgen commented on torch/_inductor/codegen/triton_utils.py:
```diff
@@ -119,6 +121,7 @@ def is_aligned(x: KernelArgType, alignment: int, include_tensor: bool) -> bool:
         for i, arg in zip(indices, args)
         if isinstance(arg, SizeArg)
         and arg.expr is not None
+        and isinstance(arg.expr, (int, sympy.Expr))
```
Why did you also include `sympy.Expr`? not symint?

unit test does not cover this case

aakhundov replied:
If I don't include sympy.Expr, symbols statically know to be equal to 1 won't be considered. I assume that the next line

```
and V.graph.sizevars.statically_known_equals(arg.expr, 1)
```

will filter out non-1 `sympy.Expr`s (including 1.0). @ezyang could you confirm if this is a valid assumption? Or should I replace with `sympy.Integer`?

ezyang replied:
So, I agree that we won't have SymInts at this point in time, in Inductor generally you're only ever dealing with int/sympy.Expr

That being said, I don't really know what the type invariants on SizeArg are, so I don't really know how to review this. Maybe I'll be satisfied with a comment here explaining the situation.

aakhundov replied:
Added a comment.

aakhundov replied:
@ezyang does this look good to you now? Thanks!

aakhundov committed:
Update on "Don't add non-integer Triton kernel arg 1 to equal_to_1"


Summary: Triton compiler adds constnat argument 1 to `equal_to_1` [only when it's an int](https://github.com/openai/triton/blob/8c5e33c77ef83e0cb99c744e58842930e602df31/python/triton/runtime/jit.py#L275). Here we restrict Inductor's `equal_to_1` in the same way.

Test Plan:

```
$ python test/inductor/test_triton_kernels.py -k test_triton_kernel_equal_to_1_float_arg
...
----------------------------------------------------------------------
Ran 1 test in 6.528s

OK

$ python test/inductor/test_triton_kernels.py -k test_triton_kernel_equal_to_1_arg
...
----------------------------------------------------------------------
Ran 2 tests in 10.142s

OK
```

cc voznesenskym penguinwu EikanWang jgong5 Guobing-Chen XiaobingSuper zhuhaozhe blzheng wenzhe-nrv jiayisunx peterbell10 ipiszy yf225 chenyang78 kadeng muchulee8 ColinPeppler amjames desertfire chauhang

[ghstack-poisoned]

aakhundov committed:
Update on "Don't add non-integer Triton kernel arg 1 to equal_to_1"


Summary: Triton compiler adds constnat argument 1 to `equal_to_1` [only when it's an int](https://github.com/openai/triton/blob/8c5e33c77ef83e0cb99c744e58842930e602df31/python/triton/runtime/jit.py#L275). Here we restrict Inductor's `equal_to_1` in the same way.

Test Plan:

```
$ python test/inductor/test_triton_kernels.py -k test_triton_kernel_equal_to_1_float_arg
...
----------------------------------------------------------------------
Ran 1 test in 6.528s

OK

$ python test/inductor/test_triton_kernels.py -k test_triton_kernel_equal_to_1_arg
...
----------------------------------------------------------------------
Ran 2 tests in 10.142s

OK
```

cc voznesenskym penguinwu EikanWang jgong5 Guobing-Chen XiaobingSuper zhuhaozhe blzheng wenzhe-nrv jiayisunx peterbell10 ipiszy yf225 chenyang78 kadeng muchulee8 ColinPeppler amjames desertfire chauhang

[ghstack-poisoned]

aakhundov committed:
Update on "Don't add non-integer Triton kernel arg 1 to equal_to_1"


Summary: Triton compiler adds constnat argument 1 to `equal_to_1` [only when it's an int](https://github.com/openai/triton/blob/8c5e33c77ef83e0cb99c744e58842930e602df31/python/triton/runtime/jit.py#L275). Here we restrict Inductor's `equal_to_1` in the same way.

Test Plan:

```
$ python test/inductor/test_triton_kernels.py -k test_triton_kernel_equal_to_1_float_arg
...
----------------------------------------------------------------------
Ran 1 test in 6.528s

OK

$ python test/inductor/test_triton_kernels.py -k test_triton_kernel_equal_to_1_arg
...
----------------------------------------------------------------------
Ran 2 tests in 10.142s

OK
```

cc voznesenskym penguinwu EikanWang jgong5 Guobing-Chen XiaobingSuper zhuhaozhe blzheng wenzhe-nrv jiayisunx peterbell10 ipiszy yf225 chenyang78 kadeng muchulee8 ColinPeppler amjames desertfire chauhang

[ghstack-poisoned]

aakhundov commented:
We discussed this offline with @oulgen and, due to `statically_known_equals` equating `sympy.Float(1.0)` and `sympy.Integer(1)`, decided to restrict possible types to `sympy.Integer` instead of the broader `sympy.Expr`.

oulgen commented on torch/_inductor/codegen/wrapper.py:
```diff
@@ -1063,7 +1063,13 @@ def define_user_defined_triton_kernel(self, kernel, configs, kwargs):
                     signature.append(SizeArg(key, arg))
-                    if arg is not None and V.graph.sizevars.statically_known_equals(arg, 1):  # type: ignore[arg-type]
+                    if (
+                        arg is not None
```
I assume you no longer need arg is not None here with the instance check

oulgen commented on torch/_inductor/codegen/triton_utils.py:
```diff
@@ -117,8 +119,11 @@ def is_aligned(x: KernelArgType, alignment: int, include_tensor: bool) -> bool:
         for i, arg in zip(indices, args)
-        if isinstance(arg, SizeArg)
-        and arg.expr is not None
+        if isinstance(arg, SizeArg) and arg.expr is not None
```
I assume you no longer need arg.expr is not None here with the instance check

oulgen commented on test/inductor/test_aot_inductor.py:
```diff
@@ -1846,6 +1847,27 @@ def forward(self, x, y):
+                out = torch.empty_like(x)
+                n_elements = x.numel()
+                add_kernel_with_scaling[(n_elements,)](
+                    x, y, out, n_elements, 1.0, BLOCK_SIZE=16
```
add dynamic here too?

aakhundov committed:
Update on "Don't add non-integer Triton kernel arg 1 to equal_to_1"


Summary: Triton compiler adds constnat argument 1 to `equal_to_1` [only when it's an int](https://github.com/openai/triton/blob/8c5e33c77ef83e0cb99c744e58842930e602df31/python/triton/runtime/jit.py#L275). Here we restrict Inductor's `equal_to_1` in the same way.

Test Plan:

```
$ python test/inductor/test_triton_kernels.py -k test_triton_kernel_equal_to_1_float_arg
...
----------------------------------------------------------------------
Ran 1 test in 6.528s

OK

$ python test/inductor/test_triton_kernels.py -k test_triton_kernel_equal_to_1_arg
...
----------------------------------------------------------------------
Ran 2 tests in 10.142s

OK
```

cc voznesenskym penguinwu EikanWang jgong5 Guobing-Chen XiaobingSuper zhuhaozhe blzheng wenzhe-nrv jiayisunx peterbell10 ipiszy yf225 chenyang78 kadeng muchulee8 ColinPeppler amjames desertfire chauhang

[ghstack-poisoned]

aakhundov committed:
Update on "Don't add non-integer Triton kernel arg 1 to equal_to_1"


Summary: Triton compiler adds constnat argument 1 to `equal_to_1` [only when it's an int](https://github.com/openai/triton/blob/8c5e33c77ef83e0cb99c744e58842930e602df31/python/triton/runtime/jit.py#L275). Here we restrict Inductor's `equal_to_1` in the same way.

Test Plan:

```
$ python test/inductor/test_triton_kernels.py -k test_triton_kernel_equal_to_1_float_arg
...
----------------------------------------------------------------------
Ran 1 test in 6.528s

OK

$ python test/inductor/test_triton_kernels.py -k test_triton_kernel_equal_to_1_arg
...
----------------------------------------------------------------------
Ran 2 tests in 10.142s

OK
```

cc voznesenskym penguinwu EikanWang jgong5 Guobing-Chen XiaobingSuper zhuhaozhe blzheng wenzhe-nrv jiayisunx peterbell10 ipiszy yf225 chenyang78 kadeng muchulee8 ColinPeppler amjames desertfire chauhang

[ghstack-poisoned]

aakhundov commented:
@pytorchbot merge
